### PR TITLE
Implement asynchronous texture uploads when using the Metal backend on iOS.

### PR DESCRIPTION
**Pull request by chinmaygarde**
This moves the Metal `GrContext` creation utilities from `GPUSurfaceMetal` into
a separate `IOSContext` object subclass. An analogue of this object was used in
the GL regime for the management of onscreen and offscreen contexts that were
not tied to the lifecycle of the `GPUSurface`. This pattern has now been
generalized for use with all backends that need a resource context
(`IOSContextGL` and `IOContextMetal`).

The platform views controller management in the `ExternalViewEmbedder` interface
implementation was repeated three times for [Metal][metal], [OpenGL](opengl) and
[Software](software) rendering. This repetition has been removed and a single
implementation present in the base `IOSSurface` and used on all platforms.
Addition of new client rendering APIs should not affect how the engine renders
into the platform view interleaving levels.

All rendering API selection logic has been moved into a single set of utilities
in `rendering_api_selection.h`. This enables the removal of a lot of code blocks
guarded by `FLUTTER_SHELL_ENABLE_METAL`. The remaining uses of this will be
removed when unified builds are enabled.

The Metal backend now also adds traces similar to the GL backend.

The `IOGLContext` has been renamed to `IOContextGL` to be more in line with the
convention used in this library.

Fixes https://github.com/flutter/flutter/issues/41827
Adds https://github.com/flutter/flutter/issues/52150

[metal]: https://github.com/flutter/engine/blob/1194ba2b218706a201c5d2c5325b55a5932546c5/shell/platform/darwin/ios/ios_surface_metal.mm#L55
[opengl]: https://github.com/flutter/engine/blob/1194ba2b218706a201c5d2c5325b55a5932546c5/shell/platform/darwin/ios/ios_surface_gl.mm#L95
[software]: https://github.com/flutter/engine/blob/1194ba2b218706a201c5d2c5325b55a5932546c5/shell/platform/darwin/ios/ios_surface_software.mm#L146

### CODE REVIEW

**Comment by chinmaygarde** [comment]
For whatever reason, I could not just update the branch on a closed PR and just re-open it. In any case, this is the same as https://github.com/flutter/engine/pull/17039 with updates:

@dnfield, Comments on your initial review and updates:

The `IOSRenderTarget` refactor that I had begun in the previous patch was indeed unrelated to this patch. I have backed that out.

> Adds more tracing

Most of the tracing is around image decompression and additions in new translation unit introduced in this commit. So the tracing is hardly unrelated to this patch. I'd prefer to keep this in.

> Refactors the way we set scale on the metal texture

This is exactly the same as before. See the [before](https://github.com/flutter/engine/blob/4f4fdbf21fa224ef33e5945cbdcf1220f7c409f0/shell/platform/darwin/ios/framework/Source/FlutterView.mm#L71) and [after](https://github.com/chinmaygarde/flutter_engine/blob/a66ec595ba7eca56922571b336cd9eec240d25a7/shell/platform/darwin/ios/framework/Source/FlutterView.mm#L57). That scaling is in the exact same spot as before.

> Removes some defaults from opacity and transaction handling on layers.

If you meant the group opacity stuff. Yes. That is unrelated and I have backed that out. That was was a premature optimization that I added initially to FlutterView and was copied over to FlutterOverlayView. Notice how it is YES in FlutterView but NO in FlutterOverlayView. Anyway, its back to how it was. I am not sure this is clearer but we can fix that later.

> In particular, some of them require manual tests to verify (e.g. you need to manually use an iOS device and rotate it and verify the rendering is correct).

This has been true of the software and OpenGL backends before and after my patch and this patch makes no attempts to address this. Even when the Metal stuff is tested in host harness, this will remain unaddressed.

> This will now throw if we're on a simulator, right?

Yes. I have fixed this and also added a flag that I can use in Metal enabled builds to test using the software backend. I have verified functionality with all backends in a single build. As before, software rendering will be defaulted to when using the simulator.

The primary change in this patch is reworking context management such that it is consistent on all platforms. Toward that end, I believe refactoring internals this way is essential because the current implementation is very OpenGL specific. For example, [a GL resource context is created in the platform view regardless of the rendering API](https://github.com/flutter/engine/blob/4f4fdbf21fa224ef33e5945cbdcf1220f7c409f0/shell/platform/darwin/ios/platform_view_ios.mm#L26). An argument could be made that another preprocessor macro will be sufficient to initialize a metal variant of that context but that does not take into account passing this variant to the `createSurface:` call or managing the device and command buffers that absent in OpenGL. So that refactor is not just happy spring cleaning but an integral part of how multiple contexts are created and managed and it is not possible to land them separately.

The parts that I do believe can be landed separately still are:
1: Cleaning up of the code repetition in IOSSurface around platform views controller.
2: Renaming ios_gl_context and ios_context_gl.

Both of these seem harmless and make the codebase a lot cleaner but I can be persuaded to back these out if you feel this is absolutely essential to landing this PR.

**Comment by dnfield** [comment]
Looks better this time - I'll try to do some manual testing with it today.

File: shell/platform/darwin/ios/ios_context.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#include "flutter/fml/macros.h"
+#include "flutter/fml/platform/darwin/scoped_nsobject.h"
+#include "flutter/shell/platform/darwin/ios/rendering_api_selection.h"
+#include "third_party/skia/include/gpu/GrContext.h"
+
+@class CALayer;
+
+namespace flutter {
+
+class IOSContext {
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_context.h]: class docstring please

File: shell/platform/darwin/ios/ios_context_metal.mm [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  main_queue_.reset([device_ newCommandQueue]);
+
+  if (!main_queue_) {
+    FML_LOG(ERROR) << "Could not create Metal command queue.";
+    return;
+  }
+
+  [main_queue_ setLabel:@"Flutter Main Queue"];
+
+  main_context_ = GrContext::MakeMetal([device_ retain], [main_queue_ retain]);
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_context_metal.mm]: This is non idiomatic, retaining in function calls.  Are you sure thats how those functions work?

File: shell/platform/darwin/ios/ios_context_software.mm [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+IOSContextSoftware::~IOSContextSoftware() = default;
+
+// |IOSContext|
+sk_sp<GrContext> IOSContextSoftware::CreateResourceContext() {
+  return nullptr;
+}
+
+// |IOSContext|
+bool IOSContextSoftware::MakeCurrent() {
+  return false;
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_context_software.mm]: Why does this return false?  You should probably add a docstring to MakeCurrent that explains what the return value means.

File: shell/platform/darwin/ios/ios_external_texture_gl.h [hunk 4]
```diff
@@ -12,34 +12,41 @@
 
 namespace flutter {
 
-class IOSExternalTextureGL : public flutter::Texture {
+class IOSExternalTextureGL final : public Texture {
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_external_texture_gl.h]: nit: for large PR's we should probably split out refactoring into a different PR.  No need to change anything here, just food for thought.

File: shell/platform/darwin/ios/ios_render_target_gl.mm [hunk 5]
```diff
@@ -12,21 +12,25 @@
...
-      valid_(false) {
+static fml::scoped_nsobject<CAEAGLLayer> CastEAGLLayer(fml::scoped_nsobject<CALayer> layer) {
+  FML_DCHECK([layer isKindOfClass:[CAEAGLLayer class]]);
+  return fml::scoped_nsobject<CAEAGLLayer>{reinterpret_cast<CAEAGLLayer*>([layer.get() retain])};
+}
+
+IOSRenderTargetGL::IOSRenderTargetGL(fml::scoped_nsobject<CALayer> layer,
+                                     fml::scoped_nsobject<EAGLContext> context,
+                                     fml::scoped_nsobject<EAGLContext> resource_context)
+    : layer_(CastEAGLLayer(std::move(layer))),
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_render_target_gl.mm]: It probably makes more sense to make this method take in a EAGLLayer and push the cast higher in the callstack.

File: shell/platform/darwin/ios/ios_render_target_gl.mm [hunk 6]
```diff
@@ -87,20 +98,23 @@
   glDiscardFramebufferEXT(GL_FRAMEBUFFER, sizeof(discards) / sizeof(GLenum), discards);
 
   glBindRenderbuffer(GL_RENDERBUFFER, colorbuffer_);
-  return [[EAGLContext currentContext] presentRenderbuffer:GL_RENDERBUFFER];
+  auto current_context = [EAGLContext currentContext];
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_render_target_gl.mm]: nit: this might be a source of slowness, querying the current context each time we present.  No need to change anything, just flagging it.

File: shell/platform/darwin/ios/ios_surface.h [hunk 7]
```diff
@@ -9,25 +9,37 @@
...
+  // |ExternalViewEmbedder|
   virtual ~IOSSurface();
 
-  virtual bool IsValid() const = 0;
+  std::shared_ptr<IOSContext> GetContext() const;
+
+  ExternalViewEmbedder* GetExternalViewEmbedderIfEnabled();
 
-  virtual bool ResourceContextMakeCurrent() = 0;
+  bool ResourceContextMakeCurrent();
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_surface.h]: docstring for return value

File: shell/platform/darwin/ios/ios_surface_gl.mm [hunk 8]
```diff
@@ -6,49 +6,49 @@
...
   if (gr_context) {
     return std::make_unique<GPUSurfaceGL>(sk_ref_sp(gr_context), this, true);
   }
   return std::make_unique<GPUSurfaceGL>(this, true);
 }
 
+// |GPUSurfaceGLDelegate|
 intptr_t IOSSurfaceGL::GLContextFBO() const {
-  return IsValid() ? render_target_->framebuffer() : GL_NONE;
+  return IsValid() ? render_target_->GetFramebuffer() : 0;
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_surface_gl.mm]: Should this be -1?  I is zero an invalid FBO?  i can't remember now if counting starts at zero or one.

File: shell/platform/darwin/ios/ios_surface_metal.mm [hunk 9]
```diff
@@ -3,110 +3,65 @@
...
-
-// |IOSSurface|
-std::unique_ptr<Surface> IOSSurfaceMetal::CreateGPUSurface(GrContext* gr_context) {
-  if (gr_context) {
-    return std::make_unique<GPUSurfaceMetal>(this, sk_ref_sp(gr_context), layer_);
+IOSSurfaceMetal::IOSSurfaceMetal(fml::scoped_nsobject<CALayer> layer,
+                                 std::shared_ptr<IOSContext> context,
+                                 FlutterPlatformViewsController* platform_views_controller)
+    : IOSSurface(std::move(context), platform_views_controller),
+      layer_(CastToMetalLayer(std::move(layer))) {
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_surface_metal.mm]: same note about casting

File: shell/platform/darwin/ios/ios_surface_metal.mm [hunk 10]
```diff
@@ -3,110 +3,65 @@
...
 
-SkCanvas* IOSSurfaceMetal::CompositeEmbeddedView(int view_id) {
-  FlutterPlatformViewsController* platform_views_controller = GetPlatformViewsController();
-  FML_CHECK(platform_views_controller != nullptr);
-  return platform_views_controller->CompositeEmbeddedView(view_id);
+// |IOSSurface|
+std::unique_ptr<Surface> IOSSurfaceMetal::CreateGPUSurface(GrContext* /* unused */) {
+  auto metal_context = CastToMetalContext(GetContext());
+
+  return std::make_unique<GPUSurfaceMetal>(this,    // Metal surface delegate
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/ios_surface_metal.mm]: Passing a raw c++ pointer inside of a unique_ptr.  Can we use weak_ptr?

File: shell/platform/darwin/ios/platform_view_ios.h [hunk 11]
```diff
@@ -17,16 +17,19 @@
...
+#include "flutter/shell/platform/darwin/ios/rendering_api_selection.h"
 
 @class FlutterViewController;
 
 namespace flutter {
 
 class PlatformViewIOS final : public PlatformView {
  public:
-  explicit PlatformViewIOS(PlatformView::Delegate& delegate, flutter::TaskRunners task_runners);
+  explicit PlatformViewIOS(PlatformView::Delegate& delegate,
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/platform_view_ios.h]: nit, explicit is usually reserved for single parameter constructors.  I'm not even sure what it means here, auto cast from `{delegate, rendering_api, task_runners}`?

File: shell/platform/darwin/ios/rendering_api_selection.h [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+enum class IOSRenderingAPI {
+  kSoftware,
+  kOpenGLES,
+  kMetal,
+};
+
+IOSRenderingAPI GetRenderingAPIForProcess();
+
+Class GetQuartzLayerClassForRenderingAPI(
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/rendering_api_selection.h]: Man, `Quartz` sounds so 2005, is that really how it is still referred to in the API?

File: shell/platform/darwin/ios/rendering_api_selection.mm [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+  if ([[[NSProcessInfo processInfo] arguments] containsObject:@"--force-metal"]) {
+    return true;
+  }
+
+  // This is just a version we picked that is easy to support and has all necessary Metal features.
+  bool ios_version_supports_metal = false;
+  if (@available(iOS 11.0, *)) {
+    ios_version_supports_metal = true;
+  }
+
```
Comment (gaaclarke) [inline shell/platform/darwin/ios/rendering_api_selection.mm]: Should there be a query of the underlying hardware?  If you have iOS 11 does that mean your phone has to support it?  We should add that info in the comment if so.

File: shell/platform/darwin/ios/ios_context_metal.mm (hunk 2 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/ios_context_metal.mm]: If it is correct, consider adding a comment here since it is non-idiomatic.

File: shell/platform/darwin/ios/rendering_api_selection.mm (hunk 13 above)
Comment (dnfield) [reply shell/platform/darwin/ios/rendering_api_selection.mm]: This is a Skia constraint.  Metal is supported from something like 8.0, but Skia only supports down to 11.0.

Is there some way we could parameterize the Skia constraint here from some constant in Skia?

File: shell/platform/darwin/ios/ios_surface.h [hunk 14]
```diff
@@ -9,25 +9,37 @@
...
+
+  // TODO(chinmaygarde): Required?
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/ios_surface.h]: This was a note to myself. The validity check is not used. This whole interface needs to be reworked. I am going to do that separately. Removed this comment.

File: shell/platform/darwin/ios/ios_surface.h (hunk 7 above)
Comment (chinmaygarde) [reply shell/platform/darwin/ios/ios_surface.h]: This is not a new method. I just removed the virtual. This whole interface needs to be reworked as the IOSContext is capable of handling this on its own. I will rework this in a later patch as it is unrelated to the task at hand.

File: shell/platform/darwin/ios/ios_surface_gl.mm (hunk 8 above)
Comment (chinmaygarde) [reply shell/platform/darwin/ios/ios_surface_gl.mm]:
```
#define GL_NONE 0
```

0 is an alias for default window-bound framebuffer on some platforms. I am not sure we we want it to be 0 as I'd prefer to default to -1 in case someone forgets to bind a framebuffer explicitly. But didn't want to change this behavior. Just didn't want to depend on OpenGL headers here but I see that we already depend on them here. So I am bringing back `GL_NONE`.

File: shell/platform/darwin/ios/ios_surface_metal.mm (hunk 9 above)
Comment (chinmaygarde) [reply shell/platform/darwin/ios/ios_surface_metal.mm]: Now all the casts are in the IOSSurface<Platform>. Do you want them higher?

File: shell/platform/darwin/ios/ios_surface_metal.mm (hunk 10 above)
Comment (chinmaygarde) [reply shell/platform/darwin/ios/ios_surface_metal.mm]: This is an existing interface that I didn't want to change.

File: shell/platform/darwin/ios/platform_view_ios.h (hunk 11 above)
Comment (chinmaygarde) [reply shell/platform/darwin/ios/platform_view_ios.h]: This was already there. But, agreed. We have not had a straight story on where/whether to use `explicit`. 

> I'm not even sure what it means here, auto cast from {delegate, rendering_api, task_runners}?

Yeah. Though I am not sure why anyone would do that. Personally, I have never found utility in the use of explicit but I can be easily swayed by an appropriate recommendation in a style guide.

File: shell/platform/darwin/ios/rendering_api_selection.h (hunk 12 above)
Comment (chinmaygarde) [reply shell/platform/darwin/ios/rendering_api_selection.h]: No. Thats just me dating myself for no reason. Changed to `GetCoreAnimationLayerClassForRenderingAPI()`.

File: shell/platform/darwin/ios/rendering_api_selection.mm (hunk 13 above)
Comment (chinmaygarde) [reply shell/platform/darwin/ios/rendering_api_selection.mm]: I believe we can go lower still but I haven't tested it yet to confirm. This constant is still WIP and this code was just moved here and not authored anew. I will tackle this separately.

File: shell/platform/darwin/ios/ios_surface_metal.mm (hunk 9 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/ios_surface_metal.mm]: Here is my thought: It's silly to cast something in a constructor because someone should have and know the requirements to make an object.  This just hides the cast better and casts are always something that should be regarded with skepticism since they can be the source of nasty bugs.

I'd say this is something similar to [demeter's law](https://en.wikipedia.org/wiki/Law_of_Demeter).  where you would prefer `int foo(Child x) => x.foo` instead of `int foo(Parent x) => x.child.foo`.

File: shell/platform/darwin/ios/rendering_api_selection.mm (hunk 13 above)
Comment (gaaclarke) [reply shell/platform/darwin/ios/rendering_api_selection.mm]: can you link in an issue so we don't lose it?

Comment (chinmaygarde) [reply shell/platform/darwin/ios/rendering_api_selection.mm]: Filed https://github.com/flutter/flutter/issues/52356 and also linked in a TODO here.

**Comment by chinmaygarde** [comment]
The LUCI error is an infra failure on the bot_update phase.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 15]
```diff
@@ -474,7 +474,8 @@
     overlay_view.get().frame = flutter_view_.get().bounds;
     overlay_view.get().autoresizingMask =
         (UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight);
-    std::unique_ptr<IOSSurface> ios_surface = [overlay_view.get() createSurface:nil];
+    std::unique_ptr<IOSSurface> ios_surface =
```
Comment (blasten) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: is it possible for an overlay to switch from gl_content to ios_context and vice versa?

Comment (chinmaygarde) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: Not sure I understand. The  IOContextGL is a subclass of IOSContext. Can you clarify?

Comment (blasten) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: Sorry - I meant if the `EnsureOverlayInitialized` can be called multiple times with the same `overlay_id`, but different values for `gl_context` and `ios_context`. If we are caching these overlays, should that be taken into account?

Comment (chinmaygarde) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: There is no more `gl_context_`, only the ios_context which should be the same reference created and held by the platform view.

Comment (blasten) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: what about `ios_context` and `gr_context`?

Comment (chinmaygarde) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: The iOS context owns the GrContexts. It owns one for on-screen rendering and another for off-screen rendering. The `gr_context` must be the same for all overlays and be the same as the on-screen context owned by the `IOSContext`.

Comment (blasten) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: I see. Thanks for the explanation!
